File: Filter/OSX/FilterDemoApp/ViewController.m
/*
	Copyright (C) 2016 Apple Inc. All Rights Reserved.
	See LICENSE.txt for this sample’s licensing information
	
	Abstract:
	View controller which registers an AUAudioUnit subclass in-process for easy development, connects sliders and text fields to its parameters, and embeds the audio unit's view into a subview. Uses SimplePlayEngine to audition the effect.
*/

#import "ViewController.h"
#import "AppDelegate.h"

#import "FilterDemoFramework.h"
#import "FilterDemo-Swift.h"
#import <CoreAudioKit/AUViewController.h>
#import "FilterDemoViewController.h"

#define kMinHertz 12.0f
#define kMaxHertz 20000.0f

@interface ViewController () {
    IBOutlet NSButton *playButton;
    
    IBOutlet NSSlider *cutoffSlider;
    IBOutlet NSSlider *resonanceSlider;
    
    IBOutlet NSTextField *cutoffTextField;
    IBOutlet NSTextField *resonanceTextField;
    
    FilterDemoViewController *auV3ViewController;
    
    SimplePlayEngine *playEngine;
    
    AUParameter *cutoffParameter;
    AUParameter *resonanceParameter;
    
    AUParameterObserverToken parameterObserverToken;
    NSArray<AUAudioUnitPreset *> *factoryPresets;
}
@property (weak) IBOutlet NSView *containerView;

-(IBAction)togglePlay:(id)sender;
-(IBAction)changedCutoff:(id)sender;
-(IBAction)changedResonance:(id)sender;

-(void)handleMenuSelection:(id)sender;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view.
    [self embedPlugInView];
    
    AudioComponentDescription desc;
    /*  Supply the correct AudioComponentDescription based on your AudioUnit type, manufacturer and creator.
     
        You need to supply matching settings in the AUAppExtension info.plist under:
         
        NSExtension
            NSExtensionAttributes
                AudioComponents
                    Item 0
                        type
                        subtype
                        manufacturer
         
         If you do not do this step, your AudioUnit will not work!!!
     */
    // MARK: AudioComponentDescription Important!
    // Ensure that you update the AudioComponentDescription for your AudioUnit type, manufacturer and creator type.
    desc.componentType = 'aufx';
    desc.componentSubType = 'f1tR';
    desc.componentManufacturer = 'Demo';
    desc.componentFlags = 0;
    desc.componentFlagsMask = 0;
    
    [AUAudioUnit registerSubclass: AUv3FilterDemo.class
           asComponentDescription: desc
                             name: @"Demo: Local AUv3"
                          version: UINT32_MAX];
    
    playEngine = [[SimplePlayEngine alloc] initWithComponentType: desc.componentType componentsFoundCallback: nil];
    [playEngine selectAudioUnitWithComponentDescription2:desc completionHandler:^{
        [self connectParametersToControls];
    }];

    [cutoffSlider sendActionOn:NSLeftMouseDraggedMask | NSLeftMouseDownMask];
    [resonanceSlider sendActionOn:NSLeftMouseDraggedMask | NSLeftMouseDownMask];
    
    [self populatePresetMenu];
}

#pragma mark -

- (void)embedPlugInView {
    NSURL *builtInPlugInURL = [[NSBundle mainBundle] builtInPlugInsURL];
    NSURL *pluginURL = [builtInPlugInURL URLByAppendingPathComponent: @"FilterDemoAppExtension.appex"];
    NSBundle *appExtensionBundle = [NSBundle bundleWithURL: pluginURL];
    
    auV3ViewController = [[FilterDemoViewController alloc] initWithNibName: @"FilterDemoViewController"
                                                                    bundle: appExtensionBundle];
    
    NSView *view = auV3ViewController.view;
    view.frame = _containerView.bounds;
    
    [_containerView addSubview: view];
    
    view.translatesAutoresizingMaskIntoConstraints = NO;
    
    NSArray *constraints = [NSLayoutConstraint constraintsWithVisualFormat: @"H:|-[view]-|"
                                                                   options:0 metrics:nil
                                                                     views:NSDictionaryOfVariableBindings(view)];
    [_containerView addConstraints: constraints];
    
    constraints = [NSLayoutConstraint constraintsWithVisualFormat: @"V:|-[view]-|"
                                                          options:0 metrics:nil
                                                            views:NSDictionaryOfVariableBindings(view)];
    [_containerView addConstraints: constraints];
}

-(void) connectParametersToControls {
    AUParameterTree *parameterTree = playEngine.testAudioUnit.parameterTree;
    
    auV3ViewController.audioUnit = (AUv3FilterDemo *)playEngine.testAudioUnit;
    cutoffParameter = [parameterTree valueForKey: @"cutoff"];
    resonanceParameter = [parameterTree valueForKey: @"resonance"];
    
    __weak ViewController *weakSelf = self;
    parameterObserverToken = [parameterTree tokenByAddingParameterObserver:^(AUParameterAddress address, AUValue value) {
        dispatch_async(dispatch_get_main_queue(), ^{
            __strong ViewController *strongSelf = weakSelf;
            
            if (address == cutoffParameter.address)
                [strongSelf updateCutoff];
            else if (address == resonanceParameter.address)
                [strongSelf updateResonance];
        });
    }];
    
    [self updateCutoff];
    [self updateResonance];
}

#pragma mark-
#pragma mark: <NSWindowDelegate>

- (void)windowWillClose:(NSNotification *)notification {
    // Main applicaiton window closing, we're done
    [playEngine stopPlaying];
    [playEngine.testAudioUnit.parameterTree removeParameterObserver:parameterObserverToken];
    
    playEngine = nil;
    auV3ViewController = nil;
}

#pragma mark-

static double logValueForNumber(double number) {
    return log(number)/log(2);
}

static double frequencyValueForSliderLocation(double location) {
    double value = powf(2, location); // (this gives us 2^0->2^9)
    value = (value - 1) / 511;        // (normalize based on rage of 2^9-1)

    // map to frequency range
    value *= (kMaxHertz - kMinHertz);
    
    return value + kMinHertz;
}

-(void) updateCutoff {
    cutoffTextField.stringValue = [cutoffParameter stringFromValue:nil];
    
    double cutoffValue = cutoffParameter.value;
    
    // normalize the value from 0-1
    double normalizedValue = ((cutoffValue - kMinHertz) / (kMaxHertz - kMinHertz));
    
    // map to 2^0 - 2^9 (slider range)
    normalizedValue = (normalizedValue * 511.0) + 1;
    
    double location = logValueForNumber(normalizedValue);
    cutoffSlider.doubleValue = location;
}

-(void) updateResonance {
    resonanceTextField.stringValue = [resonanceParameter stringFromValue: nil];
    resonanceSlider.doubleValue = resonanceParameter.value;
    
	[resonanceTextField setNeedsDisplay: YES];
    [resonanceSlider setNeedsDisplay: YES];
}

#pragma mark-
#pragma mark: Actions

-(IBAction)togglePlay:(id)sender {
    BOOL isPlaying = [playEngine togglePlay];
    
    [playButton setTitle: isPlaying ? @"Stop" : @"Play"];
}

-(IBAction)changedCutoff:(id)sender {
    if (sender == cutoffTextField)
        cutoffParameter.value = ((NSControl *)sender).doubleValue;
    else if (sender == cutoffSlider) {
        // map to frequency value
        double value = frequencyValueForSliderLocation(((NSControl *)sender).doubleValue);
        cutoffParameter.value = value;
    }
}

-(IBAction)changedResonance:(id)sender {
    if (sender == resonanceSlider || sender == resonanceTextField)
        resonanceParameter.value = ((NSControl *)sender).doubleValue;
}

#pragma mark-
#pragma mark Application Preset Menu

-(void)populatePresetMenu {
    NSApplication *app = [NSApplication sharedApplication];
    NSMenu *presetMenu = [[app.mainMenu itemWithTag:666] submenu];
    
    factoryPresets = auV3ViewController.audioUnit.factoryPresets;
    
    for (AUAudioUnitPreset *thePreset in factoryPresets) {
        NSString *keyEquivalent = @"";
        
        if (thePreset.number <= 10) {
            long keyValue = ((thePreset.number < 10) ? (long)(thePreset.number + 1) : 0);
            keyEquivalent =[NSString stringWithFormat: @"%ld", keyValue];
        }
        
        NSMenuItem *newItem = [[NSMenuItem alloc] initWithTitle:thePreset.name
                                                         action:@selector(handleMenuSelection:)
                                                  keyEquivalent:keyEquivalent];
        newItem.tag = thePreset.number;
        [presetMenu addItem:newItem];
    }
    
    AUAudioUnitPreset *currentPreset = auV3ViewController.audioUnit.currentPreset;
    [presetMenu itemAtIndex: currentPreset.number].state = NSOnState;
}

-(void)handleMenuSelection:(NSMenuItem *)sender {
    
    for (NSMenuItem *menuItem in [sender.menu itemArray]) {
        menuItem.state = NSOffState;
    }
    
    sender.state = NSOnState;
    auV3ViewController.audioUnit.currentPreset = [factoryPresets objectAtIndex:sender.tag];
}

@end

File: Filter/iOS/FilterDemoApp/ViewController.swift
/*
	Copyright (C) 2016 Apple Inc. All Rights Reserved.
	See LICENSE.txt for this sample’s licensing information
	
	Abstract:
	View controller which registers an AUAudioUnit subclass in-process for easy development, connects sliders and text fields to its parameters, and embeds the audio unit's view into a subview. Uses SimplePlayEngine to audition the effect.
*/

import UIKit
import AudioToolbox
import FilterDemoFramework

class ViewController: UIViewController {
    // MARK: Properties

	@IBOutlet var playButton: UIButton!

	@IBOutlet var cutoffSlider: UISlider!
	@IBOutlet var resonanceSlider: UISlider!
	
	@IBOutlet var cutoffTextField: UITextField!
	@IBOutlet var resonanceTextField: UITextField!

    /// Container for our custom view.
    @IBOutlet var auContainerView: UIView!
    
    static let defaultMinHertz: Double = 12.0
    static let defaultMaxHertz: Double = 20000.0
    
    let logBase = 2

	/// The audio playback engine.
	var playEngine: SimplePlayEngine!

	/// The audio unit's filter cutoff frequency parameter object.
	var cutoffParameter: AUParameter!

	/// The audio unit's filter resonance parameter object.
	var resonanceParameter: AUParameter!

	/// A token for our registration to observe parameter value changes.
	var parameterObserverToken: AUParameterObserverToken!

	/// Our plug-in's custom view controller. We embed its view into `viewContainer`.
	var filterDemoViewController: FilterDemoViewController!

    // MARK: View Life Cycle
    
	override func viewDidLoad() {
		super.viewDidLoad()
		
		// Set up the plug-in's custom view.
		embedPlugInView()
		
		// Create an audio file playback engine.
		playEngine = SimplePlayEngine(componentType: kAudioUnitType_Effect)
		
		/*
			Register the AU in-process for development/debugging.
			First, build an AudioComponentDescription matching the one in our 
            .appex's Info.plist.
		*/
        // MARK: AudioComponentDescription Important!
        // Ensure that you update the AudioComponentDescription for your AudioUnit type, manufacturer and creator type.
        var componentDescription = AudioComponentDescription()
        componentDescription.componentType = kAudioUnitType_Effect
        componentDescription.componentSubType = 0x666c7472 /*'fltr'*/
        componentDescription.componentManufacturer = 0x44656d6f /*'Demo'*/
        componentDescription.componentFlags = 0
        componentDescription.componentFlagsMask = 0
		
		/*
			Register our `AUAudioUnit` subclass, `AUv3FilterDemo`, to make it able 
            to be instantiated via its component description.
			
			Note that this registration is local to this process.
		*/
        AUAudioUnit.registerSubclass(AUv3FilterDemo.self, as: componentDescription, name:"Demo: Local FilterDemo", version: UInt32.max)

		// Instantiate and insert our audio unit effect into the chain.
		playEngine.selectAudioUnitWithComponentDescription(componentDescription) {
			// This is an asynchronous callback when complete. Finish audio unit setup.
			self.connectParametersToControls()
		}
	}
	
	/// Called from `viewDidLoad(_:)` to embed the plug-in's view into the app's view.
	func embedPlugInView() {
        /*
			Locate the app extension's bundle, in the app bundle's PlugIns
			subdirectory. Load its MainInterface storyboard, and obtain the
            `FilterDemoViewController` from that.
        */
        let builtInPlugInsURL = Bundle.main.builtInPlugInsURL!
        let pluginURL = builtInPlugInsURL.appendingPathComponent("FilterDemoAppExtension.appex")
		let appExtensionBundle = Bundle(url: pluginURL)

        let storyboard = UIStoryboard(name: "MainInterface", bundle: appExtensionBundle)
		filterDemoViewController = storyboard.instantiateInitialViewController() as! FilterDemoViewController
        
        // Present the view controller's view.
        if let view = filterDemoViewController.view {
            addChildViewController(filterDemoViewController)
            view.frame = auContainerView.bounds
            
            auContainerView.addSubview(view)
            filterDemoViewController.didMove(toParentViewController: self)
        }
	}
	
	/**
        Called after instantiating our audio unit, to find the AU's parameters and
        connect them to our controls.
    */
	func connectParametersToControls() {
		// Find our parameters by their identifiers.
        guard let parameterTree = playEngine.testAudioUnit?.parameterTree else { return }

        let audioUnit = playEngine.testAudioUnit as! AUv3FilterDemo
        filterDemoViewController.audioUnit = audioUnit
        
        cutoffParameter = parameterTree.value(forKey: "cutoff") as? AUParameter
        resonanceParameter = parameterTree.value(forKey: "resonance") as? AUParameter
        
        parameterObserverToken = parameterTree.token(byAddingParameterObserver: { [unowned self] address, value in
            /*
                This is called when one of the parameter values changes.
                
                We can only update UI from the main queue.
            */
            DispatchQueue.main.async {
                if address == self.cutoffParameter.address {
                    self.updateCutoff()
                }
                else if address == self.resonanceParameter.address {
                    self.updateResonance()
                }
            }
        })
        
        updateCutoff()
        updateResonance()
	}
    
    func logValueForNumber(_ number: Double)->Double {
        let value = log(number)/log(2);
        return value;
    }
    
    func frequencyValueForSliderLocation(_ location: Float)->Float {
        var value = pow(2, location)
        value = (value - 1)/511
        
        value *= Float(ViewController.defaultMaxHertz - ViewController.defaultMinHertz)
        
        return value + Float(ViewController.defaultMinHertz)
    }
    
	// Callbacks to update controls from parameters.
	func updateCutoff() {
		cutoffTextField.text = cutoffParameter.string(fromValue: nil)
        
        // normalize the vaue from 0-1
        let value = Double(cutoffParameter.value)
        var normalizedValue = (value - ViewController.defaultMinHertz)/(ViewController.defaultMaxHertz - ViewController.defaultMinHertz)
        
        // map to 2^0 - 2^9 (slider range)
        normalizedValue = (normalizedValue * 511.0) + 1
        
		cutoffSlider.value = Float(logValueForNumber(normalizedValue))
	}

	func updateResonance() {
		resonanceTextField.text = resonanceParameter.string(fromValue: nil)
		resonanceSlider.value = resonanceParameter.value
	}

    // MARK: IBActions

	/// Handles Play/Stop button touches.
    @IBAction func togglePlay(_ sender: AnyObject?) {
		let isPlaying = playEngine.togglePlay()

        let titleText = isPlaying ? "Stop" : "Play"

		playButton.setTitle(titleText, for: UIControlState())
	}
	
	@IBAction func changedCutoff(_ sender: AnyObject?) {
        guard sender === cutoffSlider else { return }
        
        let value = frequencyValueForSliderLocation(cutoffSlider.value)
        // Set the parameter's value from the slider's value.
        cutoffParameter.value = value
	}

	@IBAction func changedResonance(_ sender: AnyObject?) {
        guard sender === resonanceSlider else { return }

        // Set the parameter's value from the slider's value.
        resonanceParameter.value = resonanceSlider.value
	}
}
